edb/server/protocol/auth_ext/providers.py

```diff
@@ -0,0 +1,39 @@
+#
+# This source file is part of the EdgeDB open source project.
+#
+# Copyright 2022-present MagicStack Inc. and the EdgeDB authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import immutables
+
+from edb import errors
+from edb.server.config import ops
+
+from . import github
+
+SettingsMap = immutables.Map[str, ops.SettingValue]
+
+
+_providers = {
+    "github": github.GitHubProvider,
+}
+
+
+def make(db_config: SettingsMap, name: str):
+    provider_class = _providers.get(name)
+    if provider_class is None:
+        raise errors.BackendError(f"Unknown provider: {name}")
+
+    return provider_class(name, db_config)
```

Comment:
I wasn't sure where to put this logic that instantiates classes based on a map from strings to provider classes. It's lightweight enough to live in `__init__.py`, but it felt weird to import the `github` module in `__init__.py` for some reason 🤔 Any suggestions?

Reply:
> it felt weird to import the `github` module in `__init__.py` for some reason

I don't know that, but it's a common practice to do such import if you need to expose symbols from submodules onto package level.

I would import `github` within the if/case branches lazily to reduce memory usage.

Reply:
Yeah, I actually ended up using a `match` here so I can do it lazily. Does it still make sense to have a kind of separate module that is used as a dynamic constructor of another class based on some runtime information like this?